File: internal/patterns/pattern.go
package patterns

// Compiled pattern.
//
// See [StringCompile].
// See [SignedPattern.Compile].
//
// # Since 0.6.0
type PatternMinimatch struct {
	// The regular expression instance.
	//
	// # Since 0.6.0
	Re func(path string) (bool, error)
	// The original pattern string this minimatch was compiled from.
	//
	// # Since 0.6.0
	pattern string
	// The original pattern list this pattern was compiled from.
	//
	// # Since 0.6.0
	patternContext Pattern
}

func (pm PatternMinimatch) Test(path string) (bool, error) {
	return pm.Re(path)
}

// Represents a list of positive minimatch patterns.
type Pattern []string

func (pattern Pattern) Compile(options StringCompileOptions) []PatternMinimatch {
	values := []PatternMinimatch{}
	for _, p := range pattern {
		values = append(values, StringCompile(p, pattern, options))
	}
	return values
}

File: src/patterns/pattern.ts
import { stringCompile, type StringCompileOptions } from "./stringCompile.js"

/**
 * Compiled pattern.
 *
 * @see {@link stringCompile}
 * @see {@link signedPatternCompile}
 *
 * @since 0.6.0
 */
export type PatternMinimatch = {
	re: RegExp
	/**
	 * The original pattern string this minimatch was compiled from.
	 *
	 * @since 0.6.0
	 */
	pattern: string
	/**
	 * The original pattern list this pattern was compiled from.
	 *
	 * @since 0.6.0
	 */
	patternContext: Pattern
}

/**
 * Safely calls RegExp.test.
 *
 * @since 0.6.0
 */
export function patternMinimatchTest(pattern: PatternMinimatch, path: string): boolean {
	pattern.re.lastIndex = 0
	return pattern.re.test(path)
}

/**
 * Represents a list of positive minimatch patterns.
 *
 * @since 0.6.0
 */
export type Pattern = string[]

/**
 * Compiles the {@link Pattern}.
 *
 * @see {@link stringCompile}
 * @see {@link signedPatternCompile}
 *
 * @since 0.6.0
 */
export function patternCompile(
	pattern: Pattern,
	options?: StringCompileOptions,
): PatternMinimatch[] {
	return pattern.map((p, _, pattern) => stringCompile(p, pattern, options))
}
